### PR TITLE
Feature/update expired proposal

### PR DESCRIPTION
**Pull request by EricDriussi**
Added cronjob to daily bulk-change proposals status to finished whenever their closing dates are passed.

Added ability for contact person to manually change proposals status to finished given adequate criteria.

### CODE REVIEW

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java [hunk 1]
```diff
@@ -76,10 +76,57 @@
...
+                            responseCode = "200",
+                            description = "Proposal status changed to FINISHED"
+                    ),
+                    @ApiResponse(
+                            responseCode = "404",
+                            description = "Requested proposal not found"
+                    ),
+                    @ApiResponse(
+                            responseCode = "409",
+                            description = "Precondition failed, illegal status."
```
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: 409 means conflict. Precondition failed is 412.

I would rephrase it to state Conflict

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java [hunk 2]
```diff
@@ -76,10 +76,57 @@
...
+                    ),
+                    @ApiResponse(
+                            responseCode = "500",
+                            description = "Internal server error, could not fetch the user data due to a connectivity issue."
+                    ),
+            }
+    )
+    @PostMapping("/{id}/status/toFinished")
```
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: In REST APIs the PUT verb fits better when updating a resource. POST is usually used for resource creation

`/{id}/status/finished`

File: backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java [hunk 3]
```diff
@@ -76,10 +76,57 @@
...
+    @RolesAllowed("CONTACT_PERSON")
+    @ResponseStatus(HttpStatus.OK)
```
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/application/controller/ProposalApiController.java]: `200` status code is used when there is a body in the response. In this case we do not return anything so `204` is more suitable.

File: backend/src/main/java/com/huellapositiva/domain/actions/ChangeProposalStatusToFinishedAction.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    private final ProposalService proposalService;
+
+    /**
+     * This method changes the status of the proposal to FINISHED
+     *
+     * @param proposalId Id of the proposal
+     */
+    public void execute(String proposalId){
+
```
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/actions/ChangeProposalStatusToFinishedAction.java]: remove empty line

File: backend/src/main/java/com/huellapositiva/domain/actions/ChangeProposalStatusToFinishedAction.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        proposalService.changeStatusToFinished(proposalId);
+
```
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/actions/ChangeProposalStatusToFinishedAction.java]: remove empty line

File: backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java [hunk 6]
```diff
@@ -325,4 +329,41 @@ public void changeStatusToPublishedFromEnrollmentClosed(String proposalId) {
                 .name("PUBLISHED").build();
         jpaProposalRepository.updateStatusById(proposalId, jpaProposalStatus);
     }
+
+    @Scheduled(cron = "0 0 * * * * ")
```
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java]: Trim expression

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 7]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
         testData.resetData();
     }
 
+
+    @Test
+    void change_status_to_finished() throws Exception {
+        // GIVEN
+        JpaProposal publishedProp= testData.registerESALAndProposal(PUBLISHED);
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: format

What does Prop mean? Properties? Proposal? ProEvolutionSoccer? Don't be afraid of variable length, it is acceptable to call it `publishedProposal` which makes the test more readable

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 8]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
...
+        proposalService.changeStatusToFinished(publishedProp.getId());
+
+        // THEN
+        assertThat(jpaProposalRepository.findByNaturalId(publishedProp.getId()).get().getStatus().getName()).isEqualTo("finished");
+    }
+
+    @Test
+    void return_409_when_inadequate_criteria_to_change_proposal_status_to_finished() throws Exception{
+        // GIVEN
+        JpaProposal publishedProp= testData.registerESALAndProposal(CANCELLED);
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: format

File: backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java [hunk 9]
```diff
@@ -325,4 +329,41 @@ public void changeStatusToPublishedFromEnrollmentClosed(String proposalId) {
...
+        JpaProposalStatus jpaProposalStatusFinished =  JpaProposalStatus.builder()
+                .id(FINISHED.getId())
+                .name("FINISHED")
+                .build();
+
+        Calendar now = Calendar.getInstance();
+        String currentDate = String.format("%d-%d-%d", now.get(DAY_OF_MONTH), now.get(Calendar.MONTH), now.get(Calendar.YEAR));
+
+        List<JpaProposal> expiredProposals = jpaProposalRepository
+                .findExpiredProposals( new SimpleDateFormat("dd-MM-yyyy").parse(currentDate), jpaProposalStatusFinished);
```
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java]: Calendar and Date are legacy classes. Try to use the new Java Time API whenever possible.

`Date.from(Instant.now().truncatedTo(DAYS))`

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 10]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
...
+                .contentType(MediaType.APPLICATION_JSON)
+                .with(csrf())
+                .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk())
+                .andExpect(mvcResult -> assertThat(jpaProposalRepository.findByNaturalId(publishedProp.getId()).get().getStatus().getName()).isEqualTo("finished"));
+
+    }
+
+    @Test
+    void trigger_bulk_status_change_to_finished() throws Exception {
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: What does this test?

Comment (EricDriussi) [reply backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: My idea was to test that the trigger function does exactly and only what it's supposed to. 
Might be redundant, should it be removed?

File: backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java (hunk 6 above)
Comment (EricDriussi) [reply backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java]: Not sure I understand, tried with `@daily` but didn't like it.

Comment (ibaiul) [reply backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java]: Trim means removing leading and trailing spaces so `0 0 * * * * ` becomes `0 0 * * * *`

ADVANCED TIP: you should not schedule jobs at a fixed time if those jobs can be run by multiple instaces at the same time.

In production we will be running at least 2 instances of this application which means that at 00:00 every day you will have two applications loading the same entities at the same time while trying to modify them, which would lead to one potentially failing due to the locks. We do not use entity versioning at the moment but we would need to enable it in order to avoid concurrent data modification issues like corruption, staleness, consistency, ...


https://stackoverflow.com/questions/41016433/spring-scheduled-to-be-started-every-day-at-a-random-minute-between-400am-and

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java (hunk 10 above)
Comment (ibaiul) [reply backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: The test is not actually testing anything. Based on the name of the test what I would expect is:

```
// Given
Proposal 1 NOT in finished state and meets criteria to be finished
Proposal 2 NOT in finished state and does not meet criteria to be finished
...
Proposal n ...

// When
proposalService.changeExpiredProposalStatusToFinished();

// Then
Proposal 1 has been finished
Proposal 2 could not be finished
...
Proposal n ...
```


You are asserting the variable references that are not updated by any process. If you want to see if `finishedProp` proposal was changed you would need to load it from the database again.

```
JpaProposal xxx = register();
...
proposalService.changeExpiredProposalStatusToFinished();
...
xxx = findById(xxx.getId);
assert(xxx, ...);

```

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 11]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
...
+
+        testData.resetData();
+
+        JpaProposal publishedProp= testData.registerESALAndProposal(PUBLISHED);
+        publishedProp.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2099"));
+
+        testData.resetData();
+
+        JpaProposal enrollClosedProp= testData.registerESALAndProposal(ENROLLMENT_CLOSED);
+        publishedProp.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2020"));
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: This test would fail if you change the date in you computer. You should never hardcode dates in your tests

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 12]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
...
+    void trigger_bulk_status_change_to_finished() throws Exception {
+
+        // GIVEN
+        JpaProposal finishedProp= testData.registerESALAndProposal(FINISHED);
+        finishedProp.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2020"));
+
+        testData.resetData();
+
+        JpaProposal publishedProp= testData.registerESALAndProposal(PUBLISHED);
+        publishedProp.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2099"));
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: This test will break in 70 years. You should never hardcode dates in your tests. The test can only have one reason to fail, they need to be bulletproof.

`publishedProp.setClosingProposalDate(Date.from(Instant.now().plus(1, DAY)));`

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 13]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
...
+    }
+
+    @Test
+    void trigger_bulk_status_change_to_finished() throws Exception {
+
+        // GIVEN
+        JpaProposal finishedProp= testData.registerESALAndProposal(FINISHED);
+        finishedProp.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2020"));
+
+        testData.resetData();
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: I don't understand what you want to achieve with this. This line deletes the previous proposal from the database

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 14]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
...
+                .andExpect(mvcResult -> assertThat(jpaProposalRepository.findByNaturalId(publishedProp.getId()).get().getStatus().getName()).isEqualTo("finished"));
+
+    }
+
+    @Test
+    void trigger_bulk_status_change_to_finished() throws Exception {
+
+        // GIVEN
+        JpaProposal finishedProp= testData.registerESALAndProposal(FINISHED);
+        finishedProp.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2020"));
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: This line does not persist the update in the database

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 15]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
...
+
+        testData.resetData();
+
+        JpaProposal enrollClosedProp= testData.registerESALAndProposal(ENROLLMENT_CLOSED);
+        publishedProp.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2020"));
+
+        testData.resetData();
+
+        // WHEN
+        proposalService.changeExpiredProposalStatusToFinished();
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: This line is not doing anything at the moment as there are no proposals in the database

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 16]
```diff
@@ -89,6 +93,90 @@ void beforeEach() {
...
+
+        // THEN
+        assertAll(
+                ()-> assertThat(finishedProp.getStatus().getName()).isEqualTo("finished"),
+                ()-> assertThat(publishedProp.getStatus().getName()).isEqualTo("published"),
+                ()-> assertThat(enrollClosedProp.getStatus().getName()).isEqualTo("enrollment_closed")
+        );
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: You should load all the proposals you want to assert from the database to get their latest status and check if the method you are testing did change their status

File: .github/workflows/cicd.yml [hunk 17]
```diff
@@ -41,9 +41,7 @@ jobs:
       - name: Sonar
         env:
           SONAR_TOKEN: ${{ secrets.SONAR }}
-        run: |
-          BRANCH_NAME=$(echo ${GITHUB_REF} | sed 's/refs\///g' | sed 's/heads\///g' | sed 's/\/merge//g' | sed 's/pull\//PR-/g')
-          mvn sonar:sonar -Dsonar.login=${SONAR_TOKEN}  -Dsonar.branch.name=${BRANCH_NAME} --file backend/pom.xml
+        run: mvn sonar:sonar -Dsonar.login=${SONAR_TOKEN}  -Dsonar.branch.name=${GITHUB_REF} --file backend/pom.xml
```
Comment (ibaiul) [inline .github/workflows/cicd.yml]: This file should not be changed

File: backend/README.md [hunk 18]
```diff
@@ -1,4 +1,3 @@
-[![CI/CD](https://github.com/ayudadigital/huelladigital-backend/actions/workflows/cicd.yml/badge.svg?branch=develop)](https://github.com/ayudadigital/huelladigital-backend/actions/workflows/cicd.yml)
```
Comment (ibaiul) [inline backend/README.md]: This file should not be changed

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 19]
```diff
@@ -84,11 +85,94 @@
...
     @BeforeEach
     void beforeEach() {
         testData.resetData();
     }
 
+    @Test
+    void change_status_to_finished() throws Exception {
+        // GIVEN
+        JpaProposal publishedProposal= testData.registerESALAndProposal(PUBLISHED);
+        publishedProposal.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2020"));
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: Avoid hardcoding dates

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 20]
```diff
@@ -84,11 +85,94 @@
...
+
+        // WHEN
+        proposalService.changeStatusToFinished(publishedProposal.getId());
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: This test should be moved inside ProposalService

XXControllerShould is testing API requests involving HTTP request/responses

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 21]
```diff
@@ -84,11 +85,94 @@
...
+                .header(HttpHeaders.AUTHORIZATION, "Bearer " + jwtResponseDto.getAccessToken())
+                .contentType(MediaType.APPLICATION_JSON)
+                .with(csrf())
+                .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNoContent())
+                .andExpect(mvcResult -> assertThat(jpaProposalRepository.findByNaturalId(publishedProposal.getId()).get().getStatus().getName()).isEqualTo("finished"));
+    }
+
+    @Test
+    void finish_proposal_using_trigger_if_criteria_is_met() throws Exception {
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: This test belongs to ProposalServiceShould

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 22]
```diff
@@ -84,11 +85,94 @@
...
+
+        //WHEN
+        proposalService.changeExpiredProposalStatusToFinished();
+
+        //THEN
+        assertThat(jpaProposalRepository.findByNaturalId(publishedProposal.getId()).get().getStatus().getName()).isEqualTo("finished");
+    }
+
+    @Test
+    void not_finish_proposal_using_trigger_if_criteria_is_not_met() throws Exception {
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: Test belongs to ProposalServiceShould

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 23]
```diff
@@ -84,11 +85,94 @@
...
+        //GIVEN
+        JpaProposal enrollmentClosedProposal = testData.registerESALAndProposal(ENROLLMENT_CLOSED);
+        enrollmentClosedProposal.setClosingProposalDate(Date.from(Instant.now().plus(1, DAYS)));
+        jpaProposalRepository.save(enrollmentClosedProposal);
+
+        //WHEN
+        proposalService.changeExpiredProposalStatusToFinished();
+
+        //THEN
+        assertThat(jpaProposalRepository.findByNaturalId(enrollmentClosedProposal.getId()).get().getStatus().getName()).isEqualTo("enrollment_closed");
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: Extract proposal to variable to improve readability

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 24]
```diff
@@ -84,11 +85,94 @@
...
+        //GIVEN
+        JpaProposal publishedProposal = testData.registerESALAndProposal(PUBLISHED);
+        publishedProposal.setClosingProposalDate(Date.from(Instant.now().minus(1, DAYS)));
+        testData.createProposal(publishedProposal);
+
+        //WHEN
+        proposalService.changeExpiredProposalStatusToFinished();
+
+        //THEN
+        assertThat(jpaProposalRepository.findByNaturalId(publishedProposal.getId()).get().getStatus().getName()).isEqualTo("finished");
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: Extract proposal to variable to improve readability

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 25]
```diff
@@ -84,11 +85,94 @@
...
+
+
+        // WHEN + THEN
+        mvc.perform(put(FETCH_PROPOSAL_URI + publishedProposal.getId() + "/status/finished")
+                .header(HttpHeaders.AUTHORIZATION, "Bearer " + jwtResponseDto.getAccessToken())
+                .contentType(MediaType.APPLICATION_JSON)
+                .with(csrf())
+                .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isNoContent())
+                .andExpect(mvcResult -> assertThat(jpaProposalRepository.findByNaturalId(publishedProposal.getId()).get().getStatus().getName()).isEqualTo("finished"));
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: Extract proposal to variable to improve readability

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 26]
```diff
@@ -84,11 +85,94 @@
...
+
+
+        // WHEN + THEN
+        mvc.perform(put(FETCH_PROPOSAL_URI + publishedProposal.getId() + "/status/finished")
+                .header(HttpHeaders.AUTHORIZATION, "Bearer " + jwtResponseDto.getAccessToken())
+                .contentType(MediaType.APPLICATION_JSON)
+                .with(csrf())
+                .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isConflict())
+                .andExpect(mvcResult -> assertThat(jpaProposalRepository.findByNaturalId(publishedProposal.getId()).get().getStatus().getName()).isEqualTo("cancelled"));
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: Extract proposal to variable to improve readability

File: backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java [hunk 27]
```diff
@@ -84,11 +85,94 @@
...
+    void change_status_to_finished() throws Exception {
+        // GIVEN
+        JpaProposal publishedProposal= testData.registerESALAndProposal(PUBLISHED);
+        publishedProposal.setClosingProposalDate(new SimpleDateFormat("dd-MM-yyyy").parse("20-12-2020"));
+
+        // WHEN
+        proposalService.changeStatusToFinished(publishedProposal.getId());
+
+        // THEN
+        assertThat(jpaProposalRepository.findByNaturalId(publishedProposal.getId()).get().getStatus().getName()).isEqualTo("finished");
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/api/ProposalControllerShould.java]: Extract proposal to variable to improve readability

File: backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java [hunk 28]
```diff
@@ -325,4 +328,38 @@ public void changeStatusToPublishedFromEnrollmentClosed(String proposalId) {
                 .name("PUBLISHED").build();
         jpaProposalRepository.updateStatusById(proposalId, jpaProposalStatus);
     }
+
+    @Scheduled(cron = "0 ${random.int[0,59]} 0 * * ?")
+    public void changeExpiredProposalStatusToFinished() throws ParseException {
```
Comment (ibaiul) [inline backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java]: This test does not throw `ParseException` now, it is being marked as a code smell in the Sonar report. Have a look at the report, it should be easy to fix 2 out of the 3 code smells marked. You can ignore the naming convention one.

Try improving the code coverage as well to make Sonar happy :smile:

Comment (EricDriussi) [reply backend/src/main/java/com/huellapositiva/domain/service/ProposalService.java]: Crap, I forgot to check Sonar after last refactor. Sorry!

**Comment by EricDriussi** [comment]
Thought test coverage might be failing due to the Scheduler... clearly wasn't the case :rofl:

File: backend/src/test/java/com/huellapositiva/proposal/service/ProposalServiceShould.java [hunk 29]
```diff
@@ -0,0 +1,82 @@
+package com.huellapositiva.proposal.service;
```
Comment (ibaiul) [inline backend/src/test/java/com/huellapositiva/proposal/service/ProposalServiceShould.java]: The problem you are facing with the coverage is due to this test class being in the wrong package.

In our pipeline we have divided the testing phases in three: unit, integration and acceptance

These phases are executed independently and are configured in the POM file to merge their corresponding code coverage reports into a unique place. The thing is that in order to distinguish what tests belong to which phases, we configured the different testing phases to be bound to different packages.

So, unit tests are placed under `src/test/java/com/huellapositiva/unit`, integration tests under `src/test/java/com/huellapositiva/integration` and acceptance tests under `src/test/java/com/huellapositiva/api`

We could have used a different approach like naming the test classes differently depending on the test type e.g. FooTest, FooIT, FooAT 

You can see the relevant configuration in the POM and pipeline files.

To fix your issue move this test class into `src/test/java/com/huellapositiva/integration`. You can revert your experiment with the annotation and leave it as you originally planned it.

Comment (EricDriussi) [reply backend/src/test/java/com/huellapositiva/proposal/service/ProposalServiceShould.java]: Thanks a lot!

**Review by ibaiul — APPROVED** [review]
